File: Module_01_Big_O/Topic_07_Factorial_Time_O(n!)/Python/example_01.py
"""
Course title: AlgoMania
Instructor: Thiago Brito - Universo DEV
Example adapted by: Charles Fernandes de Souza
Date: October 15, 2024
"""
def factorial(max_number):
    # Initiates a loop that will traverse from 0 up to max_number-1
    for each in range(max_number):
        print(each)
        # Recursively calls the factorial function, decreasing the max_number by 1 each time
        # This recursive call generates factorial growth (O(n!)) due to the nature of multiple recursive invocations
        factorial(max_number - 1)

factorial(3)

File: Module_01_Big_O/Topic_07_Factorial_Time_O(n!)/Java/Example01/src/com/souza/charles/Example01.java
package com.souza.charles;
/**
 Course title: AlgoMania
 Instructor: Thiago Brito - Universo DEV
 Example adapted by: Charles Fernandes de Souza
 Date: October 15, 2024
 */
public class Example01 {
    public static void factorial(int maxNumber) {
        // Initiates a loop that will traverse from 0 up to maxNumber-1
        for (int each = 0; each < maxNumber; each++) {
            System.out.println(each);
            // Recursively calls the factorial function, decreasing the maxNumber by 1 each time
            // This recursive call generates factorial growth (O(n!)) due to the nature
            // of multiple recursive invocations
            factorial(maxNumber - 1);
        }
    }

    public static void main(String[] args) {
        factorial(3);
    }
}
